### ISSUE TITLE
Unsafe functions

### ISSUE THREAD

**Issue by alexeygolev** [wearegoat/bs-prosemirror#10]
We noticed that some ProseMirror functions can throw errors. Some functions actually have notes in the docs stating that they can throw. Some just throw:)

The goal of these repository is to provide bs/reason bindings with the lowest possible overhead but the goal of using ReasonML in a project is to minimise the amount of runtime errors.

With this in mind I think we'd want to add safe wrappers around unsafe functions. There are two ways of doing it — introducing new functions with `safe` suffix and introducing new functions without a suffix and adding `unsafe` suffix the bindings. Given that this library is not even published we're pretty OK with doing both.

Would love to get your opinion @tsnobip @Ben-Davis1 @SamWyld

**Comment by tsnobip** [comment]
Indeed, I've encountered that problem with `State.Transaction.maybeStep` for example.
I usually prefer to stay on the safe side, so I think it'd be better to have safe functions by default and adding an `unsafe` suffix to the normal throwing function. It's easier to see a function is wrapped looking at the `Result.t` in its signature than realizing it's unsafe and that a `safe` version of it exists too.
It's probably more idiomatic to modern OCaml anyway.

**Comment by alexeygolev** [comment]
I suggest we collect the functions that we'll need to wrap here then. We won't be able to collect them all but we'll have a good amount to start with

**Comment by Ben-Davis1** [comment]
Yeah, I definitely agree with @tsnobip (hi btw)

I realised today that the function `node` can return undefined if the depth points outside of the doc (if that makes sense). So this needs to return an option. It doesn't throw but should we include that in this list as well?

**Comment by tsnobip** [comment]
@Ben-Davis1, I guess we should make a PR to fix the bindings then and probably a PR to correct the signature in Prosemirror-model too.

**Comment by tsnobip** [comment]
the functions the docs state can throw are the following:
1. [`model.Node.replace`](https://prosemirror.net/docs/ref/#model.Node.replace), [`PM.Model.Node.replaceExn`](https://github.com/wearegoat/bs-prosemirror/blob/beebddef288e391b7395bd36ef70608f55108db4/src/PM_Model.rei#L1139) in Bs-Prosemirror
2. [`model.NodeType.createChecked`](https://prosemirror.net/docs/ref/#model.NodeType.createChecked), [`PM.Model.NodeType.createCheckedExn`](https://github.com/wearegoat/bs-prosemirror/blob/beebddef288e391b7395bd36ef70608f55108db4/src/PM_Model.rei#L666) in Bs-Prosemirror
3. [`transform.Transform.step`](https://prosemirror.net/docs/ref/#transform.Transform.step), [`PM.Transform.Transform.step`](https://github.com/wearegoat/bs-prosemirror/blob/beebddef288e391b7395bd36ef70608f55108db4/src/PM_Transform.rei#L367) in Bs-Prosemirror

some other functions throw unlike what's stated in the docs (to be completed):
1. [`transform.Transform.maybeStep`](https://prosemirror.net/docs/ref/#transform.Transform.maybeStep), [`PM.Transform.Transform.maybeStep`](https://github.com/wearegoat/bs-prosemirror/blob/beebddef288e391b7395bd36ef70608f55108db4/src/PM_Transform.rei#L373) in Bs-Prosemirror.

**Comment by alexeygolev** [comment]
@Ben-Davis1 Do you know about any other unsafe functions?
